src/storage/storage.ts

```diff
@@ -0,0 +1,634 @@
+      if (tx === null) continue;
+      yield {tx, input, index};
+    }
+  }
+
+  async addTx(tx: IHistoryTx): Promise<void> {
+    await this.store.saveTx(tx);
+  }
+
+  async processHistory(): Promise<void> {
+    await this.store.processHistory({
+      rewardLock: this.version?.reward_spend_min_blocks,
+    });
+  }
+
+  async addToken(data: ITokenData): Promise<void> {
+    await this.store.saveToken(data);
+  }
+
+  async editToken(tokenUid: string, meta: ITokenMetadata): Promise<void> {
+    this.store.editToken(tokenUid, meta);
+  }
+
+  async *getAllTokens(): AsyncGenerator<ITokenData & Partial<ITokenMetadata>> {
+    for await (const token of this.store.tokenIter()) {
+      yield token;
+    }
+  }
+
+  async *getRegisteredTokens(): AsyncGenerator<ITokenData & Partial<ITokenMetadata>> {
+    for await (const token of this.store.registeredTokenIter()) {
+      yield token;
+    }
+  }
+
+  async getToken(token?: string): Promise<(ITokenData & Partial<ITokenMetadata>)|null> {
+    return this.store.getToken(token || '00');
+  }
+
+  async registerToken(token: ITokenData): Promise<void> {
+    await this.store.registerToken(token);
+  }
+
+  async unregisterToken(tokenUid: string): Promise<void> {
+    await this.store.unregisterToken(tokenUid);
+  }
+
+  async *getAllUtxos(): AsyncGenerator<IUtxo, any, unknown> {
+    for await (const utxo of this.store.utxoIter()) {
+      yield utxo;
+    }
+  }
+
+  /**
+   * Select utxos matching the request and do not select any utxos marked for inputs.
+   *
+   * @param options Options to filter utxos and stop when the target is found.
+   *
+   * @async
+   * @generator
+   * @yields {IUtxo}
+   */
+  async *selectUtxos(options: Omit<IUtxoFilterOptions, 'reward_lock'> = {}): AsyncGenerator<IUtxo, any, unknown> {
+    const newFilter = (utxo: IUtxo): boolean => {
+      const utxoId = `${utxo.txId}:${utxo.index}`;
+      return (!this.utxosSelectedAsInput.has(utxoId)) && (options.filter_method ? options.filter_method(utxo) : true);
+    }
+
+    const newOptions: IUtxoFilterOptions = {
+      ...options,
+      filter_method: newFilter,
+    };
+    if (this.version?.reward_spend_min_blocks) {
+      newOptions.reward_lock = this.version.reward_spend_min_blocks;
+    }
+    for await (const utxo of this.store.selectUtxos(newOptions)) {
+      yield utxo;
+    }
+  }
+
+  /**
+   * Check the balance of the transaction and add inputs and outputs to match the funds and authorities.
+   * It will fail if we do not have enough funds or authorities and it will fail if we try to add too many inputs or outputs.
+   *
+   * @param tx The incomplete transaction we need to fill
+   * @param {{changeAddress?: string}} [options={}] options to use a change address.
+   *
+   * @async
+   * @returns {Promise<void>}
+   */
+  async fillTx(tx: IDataTx, { changeAddress, skipAuthorities = true, chooseInputs = true }: IFillTxOptions = {}): Promise<{inputs: IDataInput[], outputs: IDataOutput[]}> {
```

Comment:
Moved some of the method to `calculateTxBalance` (where we calculate the token balance for a given tx) and `matchTxTokensBalance` (where we generate inputs and outputs to match a given balance). The other method `matchBalanceSelection` is to reduce code duplication but it is marked as internal.

I only exposed the `calculateTxBalance` since it is useful for other parts of the wallet-lib and left the other methods as internals since they only serve to support the `fillTx` method.